KEP-4191: Split Image Filesystem add documentation (#46951)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/node-pressure-eviction.md

{{<glossary_definition term_id="node-pressure-eviction" length="short">}}</br>

{{<note>}}
The evictions signals, thresholds and other metrics that are introduced as part of

> the "Split Image" filesystem feature, which enables support for the `containerfs`

Author: Done.

> filesystem, currently require the `KubeletSeparateDiskGC` feature gate to be enabled,

Author: Done.

> filesystem, currently require the `KubeletSeparateDiskGC` feature gate to be enabled,

Author: Done.

> see [KEP-4191](https://kep.k8s.io/4191) for more details.

Author: Good to know!

I have since removed the reference to the KEP. Especially, since we also include details about different types of filesystems Kubernetes supports and the `containersfs` itself within the chapters of this document.

Hopefully, this is sufficient.

{{<glossary_definition term_id="node-pressure-eviction" length="short">}}</br>

{{<note>}}

> The eviction signals, thresholds and metrics that were introduced as part of the

Reviewer: This note reads pretty rough to me...

Also, the container runtime needs to implement this functionality. Currently this feature is only enabled for CRIO.

Author: Done.

_split image_ filesystem feature, which enables support for the `containerfs`
filesystem, require the Kubernetes v{{< skew currentVersion >}} release, which

> includes alpha versions of the feature, to enable the `KubeletSeparateDiskGC`

Reviewer: Maybe you can just mention that the feature needs to be enabled and drop mention of alpha.

Author: Done.

{{<glossary_definition term_id="node-pressure-eviction" length="short">}}</br>

{{< feature-state feature_gate_name="KubeletSeparateDiskGC" >}}

{{<note>}}
The eviction signals, thresholds and metrics that were introduced as part of the _split image_

> filesystem feature, which enables support for the `containerfs` filesystem, require both the

Reviewer 2: The feature, IMO, is "split image filesystem" instead of "split image" filesystem.
Another change I proposed was to add a manual line break.

> filesystem feature, which enables support for the `containerfs` filesystem, require both the

Reviewer 2: By the way, I was also trying to break the long sentence into shorter ones.
For documentation, we prefer simple English that are easy to parse for users.

> filesystem feature, which enables support for the `containerfs` filesystem, require both the

Author: That is fair. Give me a moment.

Kubernetes release v{{< skew currentVersion >}} to enable the `KubeletSeparateDiskGC`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) and support from

> the container runtime—currently, only CRI-O (v1.29 or higher) offers this functionality.

Author: Done.

On Linux, the kubelet uses the following eviction signals:

| Eviction Signal          | Description                                                                           |
|--------------------------|---------------------------------------------------------------------------------------|
| `memory.available`       | `memory.available` := `node.status.capacity[memory]` - `node.stats.memory.workingSet` |
| `nodefs.available`       | `nodefs.available` := `node.stats.fs.available`                                       |
| `nodefs.inodesFree`      | `nodefs.inodesFree` := `node.stats.fs.inodesFree`                                     |
| `imagefs.available`      | `imagefs.available` := `node.stats.runtime.imagefs.available`                         |
| `imagefs.inodesFree`     | `imagefs.inodesFree` := `node.stats.runtime.imagefs.inodesFree`                       |
| `containerfs.available`  | `containerfs.available` := `node.stats.runtime.containerfs.available`                 |

> | `containerfs.inodesFree` | `containerfs.inodesFree` := `node.stats.runtime.containerfs.inodesFree`               |

Reviewer 3: should we mention these are part of the split image feature gate? are they available if the gate is not on (I forget)

Author: These metrics will be available if memory serves me right, albeit I will double-check the code...

However, I didn't mention the feature gate anywhere, as we also have plans to graduate our KEP to "Beta" with this release, hopefully. So, with the flag turned on as the new default, these would be made available.

That said, I will mention the feature gate so that there is some notion of it, especially in a case should people want to switch it off, perhaps.

Reviewer 3: yeah I think mentioning it is a good idea

Author: Done.

|--------------------------|---------------------------------------------------------------------------------------|
| `memory.available`       | `memory.available` := `node.status.capacity[memory]` - `node.stats.memory.workingSet` |
| `nodefs.available`       | `nodefs.available` := `node.stats.fs.available`                                       |
| `nodefs.inodesFree`      | `nodefs.inodesFree` := `node.stats.fs.inodesFree`                                     |
| `imagefs.available`      | `imagefs.available` := `node.stats.runtime.imagefs.available`                         |
| `imagefs.inodesFree`     | `imagefs.inodesFree` := `node.stats.runtime.imagefs.inodesFree`                       |
| `containerfs.available`  | `containerfs.available` := `node.stats.runtime.containerfs.available`                 |
| `containerfs.inodesFree` | `containerfs.inodesFree` := `node.stats.runtime.containerfs.inodesFree`               |
| `pid.available`          | `pid.available` := `node.stats.rlimit.maxpid` - `node.stats.rlimit.curproc`           |

In this table, the **Description** column shows how kubelet gets the value of the

> signal. Each signal supports either a percentage or a literal value. kubelet

Author: Done.

### Eviction signals {#eviction-signals}

Eviction signals are the current state of a particular resource at a specific

> point in time. kubelet uses eviction signals to make eviction decisions by

Reviewer 3: I don't think capitalizing kubelet is incorrect here. If you'd prefer though, you could also do `The kubelet`

Author: Done.

   container images (which are the read-only layers) and container writable
   layers.

1. `containerfs`: An optional filesystem that container runtime can use to
   store everything else, similarly to what the main filesystem would (see
   `nodefs`) aside from the container images, which can then be separated (or
   split from, if you wish) such that the `imagefs` filesystem would only
   store the images (read-only layers) and nothing else.

As such, kubelet generally allows three options for container filesystems:

- Everything is on the single `nodefs`, also referred to as "rootfs" or
  simply "root", and there is no dedicated image filesystem.

- Container storage (see `nodefs`) is on a dedicated disk, and `imagefs`
  (writable and read-only layers) is separate from the root filesystem.
  This is often referred to as "Split Disk" (or "Separate Disk") filesystem.

- Container filesystem `containerfs` (same as `nodefs` plus writable
  layers) is on root and the container images (read-only layers) are
  stored on separate `imagefs`. This is often referred to as "Split Image"
  filesystem.

Currently, multiple filesystems for images and containers are not supported.

> The kubelet will attempt to auto-discover these filesystems and ignore other

Reviewer: It would probably be worth mentioning that kubelet auto-discovers this based on the container runtime configuration.

Author: Done.

> kubelet will attempt to auto-discover these filesystems and ignore other node

Reviewer 3: The kubelet here and the line below

file-backed memory on the inactive LRU list) from its calculation, as it assumes that
memory is reclaimable under pressure.

The kubelet recognizes three specific filesystem identifiers:

1. `nodefs`: The node's main filesystem, used for local disk volumes,
    emptyDir volumes not backed by memory, log storage, ephemeral storage,
    and more. For example, `nodefs` contains `/var/lib/kubelet`.

1. `imagefs`: An optional filesystem that container runtimes can use to store
   container images (which are the read-only layers) and container writable
   layers.

1. `containerfs`: An optional filesystem that container runtime can use to

>    store everything else, similarly to what the main filesystem would (see

Reviewer: ```suggestion
   store the writeable layer, similarly to what the main filesystem would (see
```

Author: Done.

>    store the writeable layer, similarly to what the main filesystem would (see

Author: Why? This pertains to the "to store" action. Why would I remove this adverb?

>    store the writeable layer, similarly to what the main filesystem would (see

Reviewer 2: You use an adjective (not an adverb) for an action here.

>    store the writeable layer, similarly to what the main filesystem would (see

Author: Are we talking about _similarly_ here?

![Screenshot from 2024-07-24 16-38-42](https://github.com/user-attachments/assets/169a2dfc-00fa-4ede-83e4-dcf714b1824a)

![Screenshot from 2024-07-24 16-40-27](https://github.com/user-attachments/assets/ebb4e24f-7f07-4345-90bd-d37922704373)

>    store the writeable layer, similarly to what the main filesystem would (see

Author: Let me rework this entire section for `containerfs` here. It needs to be improved to remove ambiguity, as per your note, about using simpler language throughout the documentation.

   `nodefs`) aside from the container images, which can then be separated (or

>    split from, if you wish) such that the `imagefs` filesystem would only

Author: Done.

   store the images (read-only layers) and nothing else.

As such, kubelet generally allows three options for container filesystems:

- Everything is on the single `nodefs`, also referred to as "rootfs" or
  simply "root", and there is no dedicated image filesystem.

- Container storage (see `nodefs`) is on a dedicated disk, and `imagefs`
  (writable and read-only layers) is separate from the root filesystem.

>   This is often referred to as "Split Disk" (or "Separate Disk") filesystem.

Author: Done.

- Container filesystem `containerfs` (same as `nodefs` plus writable
  layers) is on root and the container images (read-only layers) are

>   stored on separate `imagefs`. This is often referred to as "Split Image"

Author: Done.

  filesystem.

Currently, multiple filesystems for images and containers are not supported.

The kubelet will attempt to auto-discover these filesystems and their
current configuration directly from the underlying container runtime and,

> as such, ignore other local node filesystems; kubelet does not support

Reviewer 2: ```suggestion
as such, ignore other local node filesystems. The kubelet does not support
```

Author: Done.

> different configurations at the moment.

Reviewer 2: This sentence is confusing. I have no idea what "different configurations" means.

Author: This sentence is part of the text that was there per:

- https://github.com/kubernetes/website/commit/c184f20f76eaf58ece0ef14e025da1a4fa9d4f0d#diff-56de0c49fa24aa21138c65e0b8a9d71ab70f48159fc8184d9fee6e82793418a2R85-R86

I am not sure what precisely the original author had in mind. I can only assume, that "other" or "different" configurations refer to anything that has not been auto-discovered from the container runtime.

Author: We can drop this sentence or expand upon it. Also, perhaps @shannonxtreme recalls what was meant to be conveyed here?

Author: Done.

requests.

The kubelet sorts pods differently based on whether the node has a dedicated
`imagefs` or `containerfs` filesystem:

#### With `imagefs`

- If `nodefs` is triggering evictions, the kubelet sorts pods based on `nodefs`
  usage (`local volumes + logs of all containers`).

> - If `imagefs` is triggering evictions, the kubelet sorts pods based on the

Reviewer 3: maybe add a blurb `(assuming imagefs and containerfs are not split)` or something, as in that case writable layer sorting will happen for containerfs

Reviewer 3: oh I see. maybe the title of this section should be "Split Disk (With `imagefs` and no `containerfs`)" or something to make clearer the case

Author: Done.

requests.

The kubelet sorts pods differently based on whether the node has a dedicated
`imagefs` or `containerfs` filesystem:

> #### With `imagefs` (`nodefs` and `imagefs` filesystems are separate)

Author: Done.

- If `nodefs` triggers evictions, the kubelet sorts pods based on `nodefs`
  usage (`local volumes + logs of all containers`).

- If `imagefs` triggers evictions, the kubelet sorts pods based on the
  writable layer usage of all containers.

#### Without `imagefs` (`nodefs` and `imagefs` use the same filesystem)

- If `nodefs` triggers evictions, the kubelet sorts pods based on their
  total disk usage (`local volumes + logs and a writable layer of all containers`).

#### With `containerfs` (`imagefs` and `containerfs` have been split)

- If `containerfs` triggers evictions, the kubelet sorts pods based on
  `containerfs` usage (`local volumes + logs and a writable layer of all containers`).

- If `imagefs` triggers evictions, the kubelet sorts pods based on the
  `storage of images` rank, which represents the disk usage of a given image.

### Minimum eviction reclaim

{{<note>}}

> You currently cannot set a custom value for the `containerfs.available` metric.

Reviewer 4: ```suggestion
You cannot set a custom value for the `containerfs.available` metric.
```
(I don't think “currently” helps; you could explain that “In Kubernetes x.yy&hellip;” instead, if we have an ambition to remove this limitation).

Author: Done.

requests.

The kubelet sorts pods differently based on whether the node has a dedicated
`imagefs` or `containerfs` filesystem:

#### With `imagefs` (`nodefs` and `imagefs` filesystems are separate) {#with-imagefs}

- If `nodefs` triggers evictions, the kubelet sorts pods based on `nodefs`
  usage (`local volumes + logs of all containers`).

- If `imagefs` triggers evictions, the kubelet sorts pods based on the
  writable layer usage of all containers.

#### Without `imagefs` (`nodefs` and `imagefs` use the same filesystem) {#without-imagefs}

- If `nodefs` triggers evictions, the kubelet sorts pods based on their
  total disk usage (`local volumes + logs and a writable layer of all containers`).

#### With `containerfs` (`imagefs` and `containerfs` have been split) {#with-containersfs}

- If `containerfs` triggers evictions, the kubelet sorts pods based on
  `containerfs` usage (`local volumes + logs and a writable layer of all containers`).

> - If `imagefs` triggers evictions, the kubelet sorts pods based on the

Author: @kannon92, I derived this from the KEP itself, to quote:

> If Image filesystem experiences pressure, ranking is done as storage of images.

How this would need to be changed?

requests.

The kubelet sorts pods differently based on whether the node has a dedicated
`imagefs` or `containerfs` filesystem:

#### With `imagefs` (`nodefs` and `imagefs` filesystems are separate) {#with-imagefs}

- If `nodefs` triggers evictions, the kubelet sorts pods based on `nodefs`
  usage (`local volumes + logs of all containers`).

- If `imagefs` triggers evictions, the kubelet sorts pods based on the
  writable layer usage of all containers.

#### Without `imagefs` (`nodefs` and `imagefs` use the same filesystem) {#without-imagefs}

- If `nodefs` triggers evictions, the kubelet sorts pods based on their
  total disk usage (`local volumes + logs and a writable layer of all containers`).

> #### With `containerfs` (`imagefs` and `containerfs` have been split) {#with-containersfs}

Reviewer 3: Note: the order of these is odd. Since `with imagefs`  was previously first, it implies to me it was ordered by most customized to least, so the `with containerfs` section should be above it. I would really prefer it go `without imagefs or containerfs` `with imagefs` `with and imagefs containerfs`

Author: Sounds good.

Author: Done.
